gateway-service-impl/src/main/java/org/hypertrace/gateway/service/explore/TimeAggregationsWithGroupByRequestHandler.java

```diff
@@ -102,42 +129,143 @@ private ExploreRequest buildTimeAggregationsRequest(
   }
 
   private Filter.Builder createInClauseFilterFromGroupByResults(
-      ExploreRequest originalRequest, ExploreResponse.Builder groupByResponse) {
+      ExploreRequestContext originalRequestContext,
+      ExploreRequest originalRequest,
+      ExploreResponse.Builder groupByResponse) {
     Filter.Builder filterBuilder = Filter.newBuilder();
     filterBuilder.setOperator(Operator.AND);
     originalRequest
         .getGroupByList()
         .forEach(
             groupBy -> {
-              String groupByResultName =
-                  ExpressionReader.getSelectionResultName(groupBy).orElseThrow();
-              Set<String> inClauseValues = getInClauseValues(groupByResultName, groupByResponse);
-              filterBuilder.addChildFilter(createInClauseChildFilter(groupBy, inClauseValues));
+              Optional<Value> maybeInClauseValues =
+                  getInClauseValues(originalRequestContext, groupBy, groupByResponse);
+              maybeInClauseValues.ifPresent(
+                  inClauseValues ->
+                      filterBuilder.addChildFilter(
+                          createInClauseChildFilter(groupBy, inClauseValues)));
             });
 
     return filterBuilder;
   }
 
-  private Set<String> getInClauseValues(
-      String columnName, ExploreResponse.Builder exploreResponse) {
-    return exploreResponse.getRowBuilderList().stream()
-        .map(row -> row.getColumnsMap().get(columnName))
-        .map(Value::getString)
-        .collect(ImmutableSet.toImmutableSet());
+  private Optional<Value> getInClauseValues(
+      ExploreRequestContext originalRequestContext,
+      Expression groupBy,
+      ExploreResponse.Builder exploreResponse) {
+    String groupByResultName =
+        ExpressionReader.getAttributeIdFromAttributeSelection(groupBy).orElseThrow();
+    Map<String, AttributeMetadata> attributeMetadataMap =
+        attributeMetadataProvider.getAttributesMetadata(
+            originalRequestContext, originalRequestContext.getContext());
+    Map<String, AttributeMetadata> resultKeyToAttributeMetadataMap =
+        remapAttributeMetadataByResultName(
+            originalRequestContext.getExploreRequest(), attributeMetadataMap);
+
+    if (!resultKeyToAttributeMetadataMap.containsKey(groupByResultName)) {
+      return Optional.empty();
+    }
+
+    AttributeMetadata groupByAttributeMetadata =
+        resultKeyToAttributeMetadataMap.get(groupByResultName);
+    // RHS value of in clause filter should always be an array to apply IN filter clause on groupBy
+    // expression
+    Value.Builder valueBuilder = Value.newBuilder();
```

Comment:
can we move this out from being inline in a giant method. Ideally into its own class.

Reply:
I realize now that this is really the whole point of this method, but the generally point I think remains - it's a ~100 line method with multiple giant switches in it. Must be a way to clean that up.